File: ktor-frotagestor/src/main/kotlin/interfaces/Expense.kt
package com.frotagestor.interfaces

import kotlinx.serialization.Serializable
import kotlinx.datetime.LocalDate

@Serializable
data class Expense(
    val id: Int? = null,
    val vehicleId: Int? = null,
    val driverId: Int? = null,
    val tripId: Int? = null,
    val date: LocalDate,
    val type: String,
    val description: String? = null,
    val amount: Double,
    val liters: Double? = null,
    val pricePerLiter: Double? = null,
    val odometer: Int? = null,
    val driverName: String? = null,
    val vehiclePlate: String? = null
)

@Serializable
data class PartialExpense(
    val id: Int? = null,
    val vehicleId: Int? = null,
    val driverId: Int? = null,
    val tripId: Int? = null,
    val date: String? = null,
    val type: String? = null,
    val description: String? = null,
    val amount: Double? = null,
    val liters: Double? = null,
    val pricePerLiter: Double? = null,
    val odometer: Int? = null
)

@Serializable
data class RefuelingIndicators(
    val totalAmount: Double,
    val totalLiters: Double,
    val avgPricePerLiter: Double,
    val topDriver: TopDriver? = null,
    val topVehicleByAmount: TopVehicleAmount? = null,
    val topVehicleByLiters: TopVehicleLiters? = null,
    val lastRefueling: LastRefueling? = null
) {
    @Serializable
    data class TopDriver(val name: String, val count: Int)
    @Serializable
    data class TopVehicleAmount(val plate: String, val amount: Double)
    @Serializable
    data class TopVehicleLiters(val plate: String, val liters: Double)
    @Serializable
    data class LastRefueling(val date: String, val plate: String)
}

@Serializable
data class MaintenanceIndicators(
    val totalAmount: Double,
    val totalCount: Int,
    val mostCommonType: String,
    val topVehicleByAmount: TopVehicleAmount,
    val lastMaintenance: LastMaintenance
) {
    @Serializable
    data class TopVehicleAmount(val plate: String, val amount: Double)
    @Serializable
    data class LastMaintenance(val date: LocalDate, val plate: String)
}

@Serializable
data class ExpenseIndicators(
    val totalAmount: Double,
    val totalCount: Int,
    val mostCommonType: String? = null,
    val lastExpense: LastExpense? = null
) {
    @Serializable
    data class LastExpense(val date: LocalDate, val type: String, val description: String)
}

@Serializable
data class ExpenseReport(
    val distributions: Distributions,
    val summary: Summary
) {
    @Serializable
    data class Distributions(
        val byType: List<TypeDistribution>,
        val byVehicle: List<VehicleDistribution>,
        val byDriver: List<DriverDistribution>,
    ) {
        @Serializable
        data class TypeDistribution(val type: String, val totalAmount: Double, val totalCount: Int)
        @Serializable
        data class VehicleDistribution(val vehiclePlate: String, val totalAmount: Double, val totalCount: Int)
        @Serializable
        data class DriverDistribution(val driverName: String, val totalAmount: Double, val totalCount: Int)
    }

    @Serializable
    data class Summary(
        val totalAmount: Double,
        val totalCount: Int,
        val avgExpenseAmount: Double,
        val topExpenseType: TopExpenseType? = null,
        val topVehicleByAmount: TopVehicleAmount? = null,
        val topDriverByAmount: TopDriverAmount? = null,
        val lastExpense: LastExpense? = null
    ) {
        @Serializable
        data class TopExpenseType(val type: String, val totalAmount: Double)
        @Serializable
        data class TopVehicleAmount(val plate: String, val amount: Double)
        @Serializable
        data class TopDriverAmount(val name: String, val amount: Double)
        @Serializable
        data class LastExpense(val date: LocalDate, val type: String, val amount: Double)
    }
}

File: frontend/src/app/interfaces/expense.ts
export interface Expense {
  date: string;
  type: ExpenseType;
  amount: number;
  description: string;
  id?: number;
  vehicleId?: number | null;
  driverId?: number | null;
  vehiclePlate?: string | null;
  driverName?: string | null;
  tripId?: number | null;
  liters?: number;
  pricePerLiter?: number;
  odometer?: number;
}

export enum ExpenseType {
  COMBUSTIVEL = 'Combustivel',
  MANUTENCAO = 'Manutencao',
  ALIMENTACAO = 'Alimentacao',
  HOSPEDAGEM = 'Hospedagem',
  MULTAS = 'Multas',
  IMPOSTOS = 'Impostos',
  OUTROS = 'Outros',
}

export interface RefuelingIndicators {
  totalAmount: number;
  totalLiters: number;
  avgPricePerLiter: number;
  topDriver?: {
    name: string;
    count: number;
  };
  topVehicleByAmount?: {
    plate: string;
    amount: number;
  };
  topVehicleByLiters?: {
    plate: string;
    liters: number;
  };
  lastRefueling?: {
    date: string;
    plate: string;
  };
}

export interface MaintenanceIndicators {
  totalAmount: number;
  totalCount: number;
  mostCommonType: string;
  topVehicleByAmount: { plate: string; amount: number };
  lastMaintenance: { date: string; plate: string };
}

export interface ExpenseIndicators {
  totalAmount: number;
  totalCount: number;
  mostCommonType?: string;
  lastExpense?: {
    date: string;
    type: string;
    description: string;
  };
}

export interface ExpenseReport {
  distributions: {
    byType: { type: ExpenseType; totalAmount: number; totalCount: number }[]; // Gastos por tipo (combustível, manutenção, etc.)
    byVehicle: {
      vehiclePlate: string;
      totalAmount: number;
      totalCount: number;
    }[]; // Gastos por veículo
    byDriver: { driverName: string; totalAmount: number; totalCount: number }[]; // Gastos por motorista
    byMonth: { month: string; totalAmount: number }[]; // Gastos por mês (para linha temporal)
  };
  summary: {
    totalAmount: number; // Valor total gasto
    totalCount: number; // Número total de despesas
    avgExpenseAmount: number; // Média por despesa
    topExpenseType?: { type: ExpenseType; totalAmount: number }; // Tipo mais caro
    topVehicleByAmount?: { plate: string; amount: number }; // Veículo com maior gasto
    topDriverByAmount?: { name: string; amount: number }; // Motorista com maior gasto
    lastExpense?: { date: string; type: ExpenseType; amount: number }; // Última despesa registrada
  };
}
